### ISSUE TITLE
🤖 Code Audit: 12 potential issue(s) found

### ISSUE THREAD

**Issue by asmit25805** [studio-dots-ai/dots.tts#21]
##  Code Audit Report

> All findings are reviewed for confidence before posting.
> Please verify each finding before acting on it.

**Repository:** `rednote-hilab/dots.tts`
**Findings:** 12 issue(s) found — 🔴 1 critical · 🟠 6 high · 🟡 3 medium · 🔵 2 low

---

### 1. 🐛 Unterminated string literal causing SyntaxError

| Field | Details |
|---|---|
| **Severity** | 🔴 Critical |
| **Type** | Bug |
| **File** | `src/dots_tts/runtime.py` |
| **Location** | _prepare_inputs method |
| **Confidence** | 99% |

**Problem:**
In the _prepare_inputs method the line `raise ValueError("prompt_text requires ` is missing the closing quotation mark and the rest of the error message. This results in a syntax error, preventing the module from being imported or executed.

**Suggested Fix:**
Complete the error message and close the string literal, e.g.: `raise ValueError("prompt_text requires prompt_audio_path when prompt_text is provided")`.

---

### 2. 🔒 Potential path traversal via fid field

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Security |
| **File** | `scripts/prepare_train_jsonl_manifest.py` |
| **Location** | audio_path = (audio_dir / f"{fid}.wav").resolve() |
| **Confidence** | 92% |

**Problem:**
The script constructs the audio file path using the fid value from the CSV without sanitizing it. If a malicious fid contains path traversal components (e.g., "../"), resolve() will produce a path outside the intended audio directory, allowing the script to read arbitrary files and include them in the manifest.

**Suggested Fix:**
Validate and sanitize the fid value before using it to build a file path. For example, reject any fid containing path separators or use pathlib's PurePath to ensure the resulting path is a child of audio_dir.

---

### 3. 🔒 Tar extraction without safety checks

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Security |
| **File** | `scripts/prepare_train_jsonl_manifest.py` |
| **Location** | subprocess.run([... "tar", "-xjf", str(archive_path), ...]) |
| **Confidence** | 88% |

**Problem:**
The script extracts a tar.bz2 archive using the external 'tar' command without any validation of the archive contents. Malicious archives can contain absolute paths or symlinks that cause files to be written outside the intended extraction directory, leading to a tarbomb or privilege escalation.

**Suggested Fix:**
Replace the subprocess call with Python's tarfile module and explicitly check each member to ensure it is a regular file and that its resolved path stays within the extract_dir. Alternatively, use the '--strip-components' and '--no-same-owner' options and verify the archive before extraction.

---

### 4. 🔒 Unvalidated model path may lead to arbitrary code execution

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Security |
| **File** | `src/dots_tts/cli.py` |
| **Location** | main function - DotsTtsRuntime.from_pretrained |
| **Confidence** | 85% |

**Problem:**
The CLI accepts a user‑provided '--model-name-or-path' which is passed directly to DotsTtsRuntime.from_pretrained. If the path points to a local directory containing malicious code, the runtime may import or execute it, allowing an attacker to execute arbitrary code on the host.

**Suggested Fix:**
Validate the provided model path against an allowlist, restrict to trusted locations, or sandbox the loading process. Consider using a safe API that only loads model weights without executing arbitrary code.

---

### 5. 🐛 Function _get_initial_silence_audio_patch returns None

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `src/dots_tts/runtime_double_streaming.py` |
| **Location** | _get_initial_silence_audio_patch (missing return statement) |
| **Confidence** | 96% |

**Problem:**
The method computes a tensor called cached_patches but never returns it. Callers expecting a torch.Tensor will receive None, leading to runtime errors when the tensor is used later in the streaming pipeline.

**Suggested Fix:**
Add a return statement at the end of the function: `return cached_patches`.

---

### 6. 🔒 Arbitrary file read via unvalidated config_path

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Security |
| **File** | `src/dots_tts/config/app.py` |
| **Location** | AppConfig.from_yaml |
| **Confidence** | 92% |

**Problem:**
The `from_yaml` class method directly opens the file path provided by the caller using `Path(config_path).open` without any validation or sanitization. If an attacker can control the `config_path` argument (e.g., via a CLI argument or environment variable), they could cause the application to read any file accessible to the process, potentially exposing sensitive data or triggering further attacks.

**Suggested Fix:**
Validate the supplied `config_path` to ensure it points to an allowed location (e.g., within a predefined config directory). Reject absolute paths or paths containing '..' components, and consider using a whitelist of permitted filenames. Example:
```python
from pathlib import Path

ALLOWED_CONFIG_DIR = Path(__file__).resolve().parent.parent / "configs"

@classmethod
def from_yaml(cls, config_path: str = DEFAULT_CONFIG_PATH) -> "AppConfig":
    config_path = Path(config_path)
    # Resolve and ensure the path is inside the allowed directory
    resolved_path = (ALLOWED_CONFIG_DIR / config_path).resolve()
    if not str(resolved_path).startswith(str(ALLOWED_CONFIG_DIR)):
        raise ValueError("Config path is outside the allowed directory")
    with resolved_path.open(encoding="utf-8") as fin:
        raw_config = yaml.safe_load(fin)
    return cls.model_validate(raw_config)
```

---

### 7. 🐛 Invalid constraint on optional field 'eval_interval'

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `src/dots_tts/config/train.py` |
| **Location** | eval_interval field definition |
| **Confidence** | 94% |

**Problem:**
The field 'eval_interval' is typed as 'int | None' with a default of None, but a 'ge=1' constraint is applied. Pydantic will attempt to validate the constraint even when the value is None, leading to a validation error at model instantiation. This prevents the configuration from being created without explicitly providing an integer value.

**Suggested Fix:**
Remove the 'ge=1' constraint from the optional field or change the type to a non-optional int with a sensible default. For example: 'eval_interval: int = Field(default=100, ge=1)' or keep it optional without constraints: 'eval_interval: int | None = None'.

---

### 8. 🐛 Potential TypeError when converting configuration values to int

| Field | Details |
|---|---|
| **Severity** | 🟡 Medium |
| **Type** | Bug |
| **File** | `scripts/train_dots_tts.py` |
| **Location** | __init__ method (int conversions of config values) |
| **Confidence** | 86% |

**Problem:**
The constructor casts several configuration fields (e.g., cfg.train.max_train_steps, cfg.train.gradient_accumulation_steps, cfg.train.seed, cfg.train_data.train_audio_sample_rate) to int without checking for None. If any of these fields are omitted or set to null in the configuration, int(None) will raise a TypeError, causing the training script to crash before any work is done.

**Suggested Fix:**
Validate that each configuration value is not None before casting, or provide a default fallback. For example: `self.max_train_steps = int(cfg.train.max_train_steps) if cfg.train.max_train_steps is not None else DEFAULT_MAX_STEPS`.

---

### 9. ⚡ Silence audio patch cache is never stored

| Field | Details |
|---|---|
| **Severity** | 🟡 Medium |
| **Type** | Performance |
| **File** | `src/dots_tts/runtime_double_streaming.py` |
| **Location** | _get_initial_silence_audio_patch (cache not populated) |
| **Confidence** | 92% |

**Problem:**
The function builds a cache key and computes `cached_patches`, but it never writes the result back into the `cache` dictionary. Consequently, the expensive silence‑patch generation is performed on every call instead of being reused, hurting performance.

**Suggested Fix:**
After creating `cached_patches`, store it in the cache: `cache[key] = cached_patches` before returning the tensor.

---

### 10. 🐛 `get` method returns default even when attribute exists with the same value

| Field | Details |
|---|---|
| **Severity** | 🟡 Medium |
| **Type** | Bug |
| **File** | `src/dots_tts/config/base.py` |
| **Location** | def get(self, key: str, default=None): |
| **Confidence** | 94% |

**Problem:**
The method uses `value = getattr(self, key, default)` and immediately returns `value` if it is `default`. This means that if the attribute is present and its value equals the provided default (e.g., `None`), the method cannot distinguish between a missing attribute and an explicitly set attribute, bypassing the later `model_fields_set` check that is meant to handle this case.

**Suggested Fix:**
Rewrite the logic to first check whether the key is present in `self.model_fields_set` (or use `hasattr`) before falling back to the default. For example:
```python
def get(self, key: str, default=None):
    if key in self.model_fields_set:
        return getattr(self, key)
    return default
```

---

### 11. ⚡ Using external tar command adds overhead

| Field | Details |
|---|---|
| **Severity** | 🔵 Low |
| **Type** | Performance |
| **File** | `scripts/prepare_train_jsonl_manifest.py` |
| **Location** | subprocess.run([... "tar", "-xjf", ...]) |
| **Confidence** | 81% |

**Problem:**
Invoking the external 'tar' utility via subprocess incurs process creation overhead and blocks the Python interpreter while the extraction runs. For large archives this can increase runtime and memory usage compared to a streaming extraction in pure Python.

**Suggested Fix:**
Use the built‑in tarfile module to extract the archive in a streaming fashion, which avoids spawning a separate process and allows better integration with Python's error handling.

---

### 12. 💡 Cache language choice list to avoid recomputation

| Field | Details |
|---|---|
| **Severity** | 🔵 Low |
| **Type** | Suggestion |
| **File** | `apps/gradio/languages.py` |
| **Location** | build_language_choice_items |
| **Confidence** | 95% |

**Problem:**
The function builds a new list on each call, but the underlying data is static. Recreating the list on every invocation adds unnecessary overhead, especially if the function is called frequently (e.g., in UI rendering loops). Caching the result eliminates this repeated work.

**Suggested Fix:**
Compute the list once at module load time, e.g., LANGUAGE_CHOICES = [("不指定", ""), *[(name, code) for name, code in SUPPORTED_LANGUAGE_CODE_BY_NAME.items()]]; then have build_language_choice_items return LANGUAGE_CHOICES. Alternatively, decorate the function with @functools.lru_cache(maxsize=1) to memoize the result.

---

<details>
<summary>About this report</summary>

This report was generated using Llama 3.3 70B.
Only findings with ≥80% confidence are included.
False positives are possible — use your own judgment.

</details>
